src/tools/wasm-ctor-eval.cpp

```diff
@@ -421,11 +421,149 @@ struct CtorEvalExternalInterface : EvallingModuleRunner::ExternalInterface {
     segment.data = memory;
   }
 
+  // Serializing GC data requires more work than linear memory, because
+  // allocations have an identity, and they are created using struct.new /
+  // array.new, which we must emit in a proper location in the wasm. This
+  // affects how we serialize globals, which can contain GC data, and also, we
+  // use globals to store GC data, so overall the process of computing the
+  // globals is where most of the GC logic ends up.
+  //
+  // The general idea for handling GC data is as follows: After evaluating the
+  // code, we end up with some live allocations in the interpreter, which we
+  // need to somehow serialize into the wasm module. We will put each such live
+  // GC data item into its own "defining global", a global whose purpose is to
+  // create and store that data. Each such global is immutable, and has the
+  // exact type of the data, for simplicity. Every other reference to that GC
+  // data in the interpreter's memory can then be serialized by simply emitting
+  // a global.get of that defining global.
   void applyGlobalsToModule() {
     Builder builder(*wasm);
-    for (const auto& [name, value] : instance->globals) {
-      wasm->getGlobal(name)->init = builder.makeConstantExpression(value);
+
+    if (!wasm->features.hasGC()) {
+      // Without GC, we can simply serialize the globals in place as they are.
+      for (const auto& [name, values] : instance->globals) {
+        wasm->getGlobal(name)->init = getSerialization(values);
+      }
+      return;
     }
+
+    // We need to emit the "defining globals" of GC data before the existing
+    // globals, as the normal ones may refer to them. We do this by removing all
+    // the existing globals, and then adding them one by one, during which time
+    // we call getSerialization() for their init expressions. If their init
+    // refes to GC data, then we will allocate a defining global for that data,
+    // and refer to it. Put another way, we place the existing globals back into
+    // the module one at a time, adding their dependencies as we go.
+    auto oldGlobals = std::move(wasm->globals);
+    wasm->updateMaps();
+
+    // The process of allocating "defining globals" begins here. Clear any
+    // previous state.
+    definingGlobals.clear();
```

Comment:
I could perhaps add a function to do this explicitly instead of assuming that the globals logic is the beginning of each cycle, if you think the comments are not enough?

The reason why we have separate cycles is that each time we do a partial evaluation of a line or a function we do an update to the wasm. We need to reset the globals each time. There isn't a class whose lifetime is tied to a cycle - perhaps one could be added, but it would increase code size, and I'm not sure how/where it would make sense.

Reply:
Oh I see, I didn't realize there were multiple cycles.